compiler/rustc_hir_typeck/src/coercion.rs

```diff
@@ -582,6 +583,64 @@ impl<'f, 'tcx> Coerce<'f, 'tcx> {
 
         // Create an obligation for `Source: CoerceUnsized<Target>`.
         let cause = self.cause(self.cause.span, ObligationCauseCode::Coercion { source, target });
+        let root_obligation = Obligation::new(
+            self.tcx,
+            cause.clone(),
+            self.fcx.param_env,
+            ty::TraitRef::new(self.tcx, coerce_unsized_did, [coerce_source, coerce_target]),
+        );
+
+        // If the root `Source: CoerceUnsized<Target>` obligation can't possibly hold,
+        // we don't have to assume that this is unsizing coercion (it will always lead to an error)
+        //
+        // However, we don't want to bail early all the time, since the unholdable obligations
+        // may be interesting for diagnostics (such as trying to coerce `&T` to `&dyn Id<This = U>`)
+        // ...
+        if !self.infcx.predicate_may_hold(&root_obligation) {
+            if let Some(dyn_metadata_adt_def_id) = self.tcx.lang_items().get(LangItem::DynMetadata)
+                && let Some(metadata_type_def_id) = self.tcx.lang_items().get(LangItem::Metadata)
+            {
+                // ... so we only bail if there may be another way to convert the types.
+                //
+                // If both types are raw pointers to a (wrapper over a) trait object,
+                // this might be a cast like `*const W<dyn Trait> -> *const dyn Trait`.
+                // So it's better to bail and try that. (even if the cast is not possible, for
+                // example due to vtables not matching, cast diagnostic will likely still be better)
+                self.probe(|_| {
+                    use traits::TraitEngineExt;
+                    let mut fulfill_cx = <dyn TraitEngine<'_, ScrubbedTraitError<'_>>>::new(self);
+                    let mut normalize = |ty| {
+                        self.at(&cause, self.fcx.param_env)
+                            .structurally_normalize(ty, &mut *fulfill_cx)
+                    };
+
+                    // N.B. use `target`, not `coerce_target` (the latter is a var)
```

Comment:
Can you clarify in a comment what this is doign? also, i think the "kind match + normalize projection + check its adt kind" can be uplifted into a closure, not just the normalize part.

Reply:
er, actually maybe just move that comment from above down here?